libvast/src/table_slice.cpp

```diff
@@ -110,12 +91,32 @@ caf::error table_slice::deserialize_ptr(caf::deserializer& source,
     ptr.reset();
     return caf::none;
   }
-  ptr = make_ptr(std::move(layout), source.context()->system(), impl_id);
+  ptr = make_table_slice(std::move(layout), source.context()->system(),
+                         impl_id);
   if (!ptr)
     return ec::invalid_table_slice_type;
   return ptr->deserialize(source);
 }
 
+table_slice_ptr make_table_slice(record_type layout, caf::actor_system& sys,
+                                 caf::atom_value impl) {
+  if (impl == caf::atom("DEFAULT")) {
+    auto ptr = caf::make_counted<default_table_slice>(std::move(layout));
+    return ptr;
+  }
+  using generic_fun = caf::runtime_settings_map::generic_function_pointer;
+  using factory_fun = table_slice_ptr (*)(record_type);
+  auto val = sys.runtime_settings().get(impl);
+  if (!caf::holds_alternative<generic_fun>(val)) {
+    VAST_ERROR_ANON("table_slice",
+                    "has no factory function for implementation key",
+                    impl);
+    return nullptr;
+  }
+  auto fun = reinterpret_cast<factory_fun>(caf::get<generic_fun>(val));
```

Comment:
I don't think I understand this code. Why does this have to be solved through `runtime_settings`? Why is overloading a virtual function not enough?

Reply:
(It sounds like your question is orthogonal to this PR, which merely shuffled code around.)

> Why does this have to be solved through `runtime_settings`?

Because we store the factory function in the actor system. Only the factory knows how to create concrete table slice types, and we want it to be extensible by other applications.

Reply:
I'm aware that my question touches on a larger subject, but I still think it is quite relevant here, because this change is about the API, and that is what I want to discuss.

After looking at the code for quite a while I think I understand the reasoning behind this design. The atom is used as a handle to the factory so serialized slices can be read.

What was strange to me is that a reference to the actor system is necessary here, could the map not be a static member of `table_slice` and expose a `register_concrete_factory()` function?